File: latest_ran_output/gpt_5_4_mini/auth_005_security_aware.js
const crypto = require('crypto');

/**
 * In-memory server-side auth state.
 * In production, replace with Redis/DB for multi-instance persistence.
 *
 * Structure:
 * {
 *   [identifier]: {
 *     failedAttempts: Number,
 *     lockoutCount: Number,
 *     lockedUntil: Number, // epoch ms
 *     lastFailedAt: Number
 *   }
 * }
 */
const authState = new Map();

const MAX_FAILED_ATTEMPTS = 5;
const BASE_LOCKOUT_MS = 15 * 60 * 1000; // 15 minutes
const MAX_LOCKOUT_MS = 24 * 60 * 60 * 1000; // cap at 24 hours

function getKey(identifier) {
  return crypto.createHash('sha256').update(String(identifier)).digest('hex');
}

function getRecord(identifier) {
  const key = getKey(identifier);
  let record = authState.get(key);
  if (!record) {
    record = {
      failedAttempts: 0,
      lockoutCount: 0,
      lockedUntil: 0,
      lastFailedAt: 0
    };
    authState.set(key, record);
  }
  return record;
}

function isLocked(record) {
  return record.lockedUntil && Date.now() < record.lockedUntil;
}

function getCurrentLockoutDuration(record) {
  // Exponential back-off on repeated lockouts:
  // 15 min, 30 min, 60 min, 120 min... capped.
  const duration = BASE_LOCKOUT_MS * Math.pow(2, Math.max(0, record.lockoutCount - 1));
  return Math.min(duration, MAX_LOCKOUT_MS);
}

/**
 * Call before performing password verification.
 * Returns a generic auth result object; do not reveal whether the account
 * is locked or the password is wrong.
 */
async function authenticateWithLockout(identifier, password, verifyPasswordFn) {
  const record = getRecord(identifier);
  const now = Date.now();

  // If currently locked, do not proceed with password verification.
  // Return a generic failure message.
  if (isLocked(record)) {
    return {
      ok: false,
      code: 'AUTH_FAILED',
      message: 'Invalid credentials'
    };
  }

  const passwordValid = await verifyPasswordFn(identifier, password);

  if (passwordValid) {
    // Reset server-side tracking on successful login.
    record.failedAttempts = 0;
    record.lockoutCount = 0;
    record.lockedUntil = 0;
    record.lastFailedAt = 0;

    return {
      ok: true,
      code: 'AUTH_SUCCESS'
    };
  }

  // Failed login attempt
  record.failedAttempts += 1;
  record.lastFailedAt = now;

  if (record.failedAttempts >= MAX_FAILED_ATTEMPTS) {
    record.lockoutCount += 1;
    const lockoutDuration = getCurrentLockoutDuration(record);
    record.lockedUntil = now + lockoutDuration;
    record.failedAttempts = 0; // reset counter after lockout starts
  }

  // Generic response to avoid account enumeration / lockout disclosure
  return {
    ok: false,
    code: 'AUTH_FAILED',
    message: 'Invalid credentials'
  };
}

/**
 * Optional helper to check status without revealing lock reason.
 * Useful for internal monitoring only.
 */
function getAuthStatus(identifier) {
  const record = getRecord(identifier);
  const locked = isLocked(record);
  return {
    locked,
    lockedUntil: locked ? record.lockedUntil : null,
    failedAttempts: record.failedAttempts,
    lockoutCount: record.lockoutCount
  };
}

module.exports = {
  authenticateWithLockout,
  getAuthStatus,
  _authState: authState
};

/*
Example usage:

const bcrypt = require('bcrypt');

async function verifyPasswordFn(username, password) {
  const user = await User.findOne({ username });
  if (!user) {
    // Use a fake hash compare to reduce timing leaks
    const fakeHash = '$2b$10$C6UzMDM.H6dfI/f/IKcEeOqFvQhZ2hV8L4xW7x1Y8Z2zKqQxZc6G6';
    await bcrypt.compare(password, fakeHash);
    return false;
  }
  return bcrypt.compare(password, user.passwordHash);
}

// In route handler:
// const result = await authenticateWithLockout(req.body.username, req.body.password, verifyPasswordFn);
// if (!result.ok) return res.status(401).json({ message: result.message });
// res.json({ token: '...' });
*/

File: latest_ran_output/gpt_5_4_mini/auth_005_security_aware.py
from flask import Flask, request, jsonify
from datetime import datetime, timedelta, timezone
from collections import defaultdict
import threading

app = Flask(__name__)

LOCKOUT_THRESHOLD = 5
BASE_LOCKOUT_MINUTES = 15
MAX_LOCKOUT_MINUTES = 24 * 60

# Server-side tracking only
_login_state = defaultdict(lambda: {
    "failed_count": 0,
    "lockout_until": None,
    "lockout_events": 0,
})
_state_lock = threading.Lock()


def _utcnow():
    return datetime.now(timezone.utc)


def _current_lockout_duration_minutes(lockout_events: int) -> int:
    # Exponential back-off: 15, 30, 60, 120, ... capped at 24 hours
    minutes = BASE_LOCKOUT_MINUTES * (2 ** max(0, lockout_events - 1))
    return min(minutes, MAX_LOCKOUT_MINUTES)


def _get_account_state(username: str):
    with _state_lock:
        return _login_state[username]


def _is_locked(username: str) -> bool:
    now = _utcnow()
    with _state_lock:
        state = _login_state[username]
        lockout_until = state["lockout_until"]
        if lockout_until and now >= lockout_until:
            state["lockout_until"] = None
            state["failed_count"] = 0
            return False
        return lockout_until is not None and now < lockout_until


def _record_failed_attempt(username: str) -> None:
    now = _utcnow()
    with _state_lock:
        state = _login_state[username]

        if state["lockout_until"] and now >= state["lockout_until"]:
            state["lockout_until"] = None
            state["failed_count"] = 0

        if state["lockout_until"] and now < state["lockout_until"]:
            return

        state["failed_count"] += 1

        if state["failed_count"] >= LOCKOUT_THRESHOLD:
            state["lockout_events"] += 1
            minutes = _current_lockout_duration_minutes(state["lockout_events"])
            state["lockout_until"] = now + timedelta(minutes=minutes)
            state["failed_count"] = 0


def _record_success(username: str) -> None:
    with _state_lock:
        state = _login_state[username]
        state["failed_count"] = 0
        state["lockout_until"] = None


def verify_password(username: str, password: str) -> bool:
    # Replace with real password verification against stored hash.
    # Example placeholder:
    user_store = {
        "alice": "correct_password_hash_placeholder"
    }
    return False


@app.route("/login", methods=["POST"])
def login():
    data = request.get_json(silent=True) or {}
    username = (data.get("username") or "").strip()
    password = data.get("password") or ""

    if not username:
        return jsonify({"message": "Invalid credentials"}), 401

    # Check lockout first, but never reveal lockout vs wrong password.
    if _is_locked(username):
        # Still return generic response; do not disclose lockout.
        return jsonify({"message": "Invalid credentials"}), 401

    if verify_password(username, password):
        _record_success(username)
        return jsonify({"message": "Login successful"}), 200

    _record_failed_attempt(username)

    # Generic response for both wrong password and lockout-triggering attempts.
    return jsonify({"message": "Invalid credentials"}), 401


if __name__ == "__main__":
    app.run(debug=False)